### PR TITLE
Hot rod dialect v5

### PR DESCRIPTION
**Pull request by Sanne**
Rebased to accommodate for the SPI changes and improved the association storage strategies, as recently added tests highlighted a bug in the association kind detection.

 - https://hibernate.atlassian.net/browse/OGM-1160
 - https://hibernate.atlassian.net/browse/OGM-1161
 - https://hibernate.atlassian.net/browse/OGM-1163
 - https://hibernate.atlassian.net/browse/OGM-1164
 - https://hibernate.atlassian.net/browse/OGM-353

### CODE REVIEW

File: infinispan-remote/src/test/java/org/hibernate/ogm/datastore/infinispanremote/test/sequences/SequenceAtomicIncrementTest.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+ *
+ * By default it's not running much of a stress test as we want the tests to give
+ * quick feedback; Set the ogm.runFullStressTests System property to have it run
+ * for several minutes.
+ * Careful when enlarging constants for longer tests to also allow for more memory
+ * as we need to track and compare all generated results at the end.
+ *
+ * @author Sanne Grinovero
+ */
+public class SequenceAtomicIncrementTest {
```
Comment (DavideD) [inline infinispan-remote/src/test/java/org/hibernate/ogm/datastore/infinispanremote/test/sequences/SequenceAtomicIncrementTest.java]: I can create a Job on CI for the longer test

**Comment by DavideD** [comment]
Could you refresh a moment my memory about the two configuration files?
hotrod-server-singleton.xml
hotrod-client-testingconfiguration.properties

File: infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/VersionedTuple.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	public long getVersion() {
+		return version;
+	}
+
+	public void setVersion(long version) {
+		this.version = version;
+	}
+
+	public boolean isBeingInserted() {
+		return beingInserted;
```
Comment (DavideD) [inline infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/VersionedTuple.java]: This is not needed anymore, isn't it?

**Comment by DavideD** [comment]
Merged https://github.com/hibernate/hibernate-ogm/pull/781

Thanks, it's great to see this finally merged

File: infinispan-remote/src/test/java/org/hibernate/ogm/datastore/infinispanremote/utils/HotrodServerLifecycle.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+			printEvent( "Test suite end detected" );
+			shutdownServer();
+		}
+	}
+
+	@Override
+	public void testFinished(Description description) throws Exception {
+		if ( server != null ) {
+			// I'd like to do this, but in some cases currently OGM's tests
+			// are bit always fully independent in the same class.
```
Comment (Sanne) [inline infinispan-remote/src/test/java/org/hibernate/ogm/datastore/infinispanremote/utils/HotrodServerLifecycle.java]: Right, some tests will fail if I clean up the database between them.

File: infinispan-remote/src/test/java/org/hibernate/ogm/datastore/infinispanremote/test/sequences/SequenceAtomicIncrementTest.java (hunk 1 above)
Comment (Sanne) [inline infinispan-remote/src/test/java/org/hibernate/ogm/datastore/infinispanremote/test/sequences/SequenceAtomicIncrementTest.java]: I'm not sure it's worth it. The semantics have been carefully inspected by people on the Infinispan team, if there are failures *they* should detect them with their tests.
Also it's not meant to be a long term solution; we're discussing better alternative either using a custom task which we can deploy, or to have Hot Rod expose a sequence functionality.

**Pull request merged**